File: algorithms/318. Maximum Product of Word Lengths/solution.py
class Solution(object):
    def maxProduct(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        d = collections.defaultdict(int)
        for w in words:
            mask = sum(1 << (ord(c) - 97) for c in set(w))
            d[mask] = max(d[mask], len(w))
        return max([d[i] * d[j] for i in d for j in d if not (i & j)] or [0])

File: algorithms/318. Maximum Product of Word Lengths/solution.cc
class Solution {
public:
    int maxProduct(vector<string>& words) {
        int n = words.size(), mp = 0;
        vector<int> bitmask(n);
        for ( int i = 0; i < n; ++i ) {
            for ( int j = 0; j < words[i].size(); ++j ) {
                bitmask[i] |= 1 << (words[i][j] - 'a');
            }
        }
        for ( int i = 0; i < n; ++i ) {
            for ( int j = i + 1; j < n; j++ ) {
                if ( (bitmask[i] & bitmask[j]) == 0 ) {
                    mp = max(mp, int(words[i].size() * words[j].size()));
                }
            }
        }
        return mp;
    }
};

// 174 / 174 test cases passed.
// Status: Accepted
// Runtime: 45 ms
// beats 86.98 %
